File: clients/python/llm_chat/protocol.py
"""Async client for the manager's typed ``/chat`` WebSocket protocol.

Protocol (per message):
    client → ``{"type":"q","id":<id>,"text":<text>}``
    manager → ``{"type":"ack","id","seq"}``           (receipt)
    manager → ``{"type":"a","id","seq","text",...}``  (the answer)
    client → ``{"type":"confirm","seq":<seq>}``       (we got it)
    manager → ``{"type":"err",...}``                  (on failure)

`ChatClient` keeps ONE connection (and therefore one backend session, so claude
retains conversation context across `ask()` calls). If the socket drops it
transparently re-authenticates and reconnects on the next `ask()`.
"""

from __future__ import annotations

import asyncio
import json
import logging
import time
from dataclasses import dataclass
from typing import Awaitable, Callable

import websockets

from .errors import AnswerTimeout, ManagerUnavailable, ProtocolError

log = logging.getLogger("llm_chat.protocol")

# A callable that returns a fresh access token (used for connect + reconnect).
TokenProvider = Callable[[], Awaitable[str]] | Callable[[], str]


@dataclass(frozen=True)
class Answer:
    """A settled answer for one question."""

    text: str
    seq: int
    id: str
    time_in: str | None = None
    time_out: str | None = None

    @property
    def latency_s(self) -> float | None:
        if not (self.time_in and self.time_out):
            return None
        try:
            from datetime import datetime

            fmt = "%Y-%m-%dT%H:%M:%S.%f%z"
            return (
                datetime.strptime(self.time_out, fmt) - datetime.strptime(self.time_in, fmt)
            ).total_seconds()
        except (ValueError, TypeError):
            return None


async def _call_token_provider(provider: TokenProvider) -> str:
    result = provider()
    if asyncio.iscoroutine(result):
        return await result
    return result  # type: ignore[return-value]


class ChatClient:
    """A persistent `/chat` session. Use as an async context manager."""

    def __init__(
        self,
        manager_url: str,
        token_provider: TokenProvider,
        *,
        max_reconnects: int = 1,
    ) -> None:
        self._url = manager_url
        self._token_provider = token_provider
        self._max_reconnects = max_reconnects
        self._ws: websockets.WebSocketClientProtocol | None = None
        self._counter = 0
        self.session_id: str | None = None

    async def __aenter__(self) -> "ChatClient":
        await self.connect()
        return self

    async def __aexit__(self, *exc) -> None:
        await self.close()

    async def connect(self) -> None:
        """Open the WebSocket and drain the manager's ``initialized`` frame.

        Raises:
            ManagerUnavailable: the connection could not be established.
        """
        token = await _call_token_provider(self._token_provider)
        try:
            self._ws = await websockets.connect(
                self._url,
                additional_headers=[("Authorization", f"Bearer {token}")],
                max_size=None,
                open_timeout=15,
            )
        except (OSError, websockets.WebSocketException, asyncio.TimeoutError) as e:
            raise ManagerUnavailable(f"could not connect to {self._url}: {e}") from e

        # The manager sends an `initialized` frame right away; capture the sid.
        try:
            raw = await asyncio.wait_for(self._ws.recv(), timeout=10)
            first = json.loads(raw)
            if first.get("type") == "initialized":
                self.session_id = first.get("sid")
                log.debug("connected sid=%s backendPort=%s", self.session_id, first.get("backendPort"))
            else:
                # Not fatal — push it back conceptually by handling in ask(); but
                # the protocol always leads with `initialized`, so just log.
                log.warning("expected 'initialized', got %r", first.get("type"))
        except (asyncio.TimeoutError, websockets.WebSocketException, ValueError) as e:
            await self.close()
            raise ManagerUnavailable(f"no 'initialized' frame from {self._url}: {e}") from e

    async def close(self) -> None:
        if self._ws is not None:
            try:
                await self._ws.close()
            except Exception:  # noqa: BLE001 — close must never raise
                pass
            self._ws = None

    @property
    def connected(self) -> bool:
        # We don't probe a library-specific `.closed`/`.state` (it moved across
        # websockets versions). `close()` and the drop handler null `_ws`, and
        # ask() catches ConnectionClosed and reconnects — so this is enough.
        return self._ws is not None

    async def ask(self, text: str, *, timeout: float = 120.0) -> Answer:
        """Send a question and return its settled `Answer`.

        Transparently reconnects (up to ``max_reconnects``) if the socket has
        dropped. Raises `AnswerTimeout`, `ProtocolError`, or `ManagerUnavailable`.
        """
        attempts = 0
        while True:
            if not self.connected:
                await self.connect()
            self._counter += 1
            msg_id = f"m{self._counter}"
            try:
                await self._ws.send(json.dumps({"type": "q", "id": msg_id, "text": text}))
                return await self._await_answer(msg_id, time.monotonic() + timeout)
            except websockets.ConnectionClosed as e:
                await self.close()
                attempts += 1
                if attempts > self._max_reconnects:
                    raise ManagerUnavailable(f"connection closed during ask(): {e}") from e
                log.warning("connection dropped, reconnecting (attempt %d)", attempts)
                # loop: reconnect and retry the same question

    async def _await_answer(self, msg_id: str, deadline: float) -> Answer:
        assert self._ws is not None
        while True:
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise AnswerTimeout(f"no answer for {msg_id} within the timeout")
            try:
                raw = await asyncio.wait_for(self._ws.recv(), timeout=remaining)
            except asyncio.TimeoutError:
                raise AnswerTimeout(f"no answer for {msg_id} within the timeout") from None
            try:
                msg = json.loads(raw)
            except ValueError as e:
                raise ProtocolError(f"manager sent non-JSON frame: {raw!r}") from e
            mtype = msg.get("type")
            if mtype == "a" and msg.get("id") == msg_id:
                await self._ws.send(json.dumps({"type": "confirm", "seq": msg["seq"]}))
                return Answer(
                    text=msg.get("text", ""),
                    seq=msg.get("seq", -1),
                    id=msg.get("id", msg_id),
                    time_in=msg.get("timeIn"),
                    time_out=msg.get("timeOut"),
                )
            if mtype == "err":
                raise ProtocolError(msg.get("text", "manager returned an error"))
            # `initialized`, `ack`, or an unrelated `a` → keep waiting.
            log.debug("skip frame type=%s id=%s", mtype, msg.get("id"))

File: clients/rust/src/protocol.rs
//! Async client for the manager's typed `/chat` WebSocket. Port of `protocol.py`.
//!
//! Protocol (per message):
//!   client  -> {"type":"q","id":<id>,"text":<text>}
//!   manager -> {"type":"ack","id","seq"}            (receipt)
//!   manager -> {"type":"a","id","seq","text",...}   (the answer)
//!   client  -> {"type":"confirm","seq":<seq>}       (we got it)
//!   manager -> {"type":"err",...}                   (on failure)
//!
//! `ChatClient` keeps ONE connection (one backend session, so claude retains
//! context across `ask()` calls). If the socket drops it transparently
//! reconnects on the next `ask()`.

use std::sync::Arc;
use std::time::{Duration, Instant};

use futures_util::{SinkExt, StreamExt};
use serde_json::json;
use tokio::net::TcpStream;
use tokio_tungstenite::tungstenite::client::IntoClientRequest;
use tokio_tungstenite::tungstenite::protocol::WebSocketConfig;
use tokio_tungstenite::tungstenite::Message;
use tokio_tungstenite::{connect_async_with_config, MaybeTlsStream, WebSocketStream};

use crate::errors::{Error, Result};

type Ws = WebSocketStream<MaybeTlsStream<TcpStream>>;

/// A callable returning a fresh access token (used for connect + reconnect).
/// Blocking (it does sync HTTP), so it is invoked via `spawn_blocking`.
pub type TokenProvider = Arc<dyn Fn() -> Result<String> + Send + Sync>;

/// A settled answer for one question.
#[derive(Debug, Clone)]
pub struct Answer {
    pub text: String,
    pub seq: i64,
    pub id: String,
    pub time_in: Option<String>,
    pub time_out: Option<String>,
    pub latency_ms: Option<i64>,
}

impl Answer {
    /// Latency in seconds. Prefers the manager's `latencyMs` (computed from the
    /// same timeIn/timeOut the Python client parses), so the displayed number
    /// matches without a date-parsing dependency.
    pub fn latency_s(&self) -> Option<f64> {
        self.latency_ms.filter(|&ms| ms >= 0).map(|ms| ms as f64 / 1000.0)
    }
}

enum Awaited {
    Answer(Answer),
    Closed,
}

pub struct ChatClient {
    url: String,
    token_provider: TokenProvider,
    max_reconnects: u32,
    ws: Option<Ws>,
    counter: u64,
    pub session_id: Option<String>,
}

impl ChatClient {
    pub fn new(manager_url: &str, token_provider: TokenProvider) -> Self {
        ChatClient {
            url: manager_url.to_string(),
            token_provider,
            max_reconnects: 1,
            ws: None,
            counter: 0,
            session_id: None,
        }
    }

    pub fn connected(&self) -> bool {
        self.ws.is_some()
    }

    /// Re-mint (machine) or refresh (human) the access token — used by display
    /// surfaces like the REPL's `/status`, which decode its claims.
    pub async fn current_token(&self) -> Result<String> {
        self.fetch_token().await
    }

    /// Request THIS user's own usage and return the manager's `usage` reply
    /// object. Sends `{"type":"usage"}` and awaits the `usage` frame, skipping
    /// any interleaved ack/answer frames.
    pub async fn usage(&mut self, timeout: Duration) -> Result<serde_json::Value> {
        if !self.connected() {
            self.connect().await?;
        }
        let req = json!({"type": "usage"}).to_string();
        match self.ws.as_mut() {
            Some(ws) => ws
                .send(Message::Text(req))
                .await
                .map_err(|e| Error::ManagerUnavailable(format!("usage send failed: {e}")))?,
            None => return Err(Error::ManagerUnavailable("not connected".into())),
        }
        let deadline = Instant::now() + timeout;
        loop {
            let remaining = deadline.saturating_duration_since(Instant::now());
            if remaining.is_zero() {
                return Err(Error::AnswerTimeout("no usage reply within timeout".into()));
            }
            let ws = match self.ws.as_mut() {
                Some(w) => w,
                None => return Err(Error::ManagerUnavailable("connection closed".into())),
            };
            let frame = match tokio::time::timeout(remaining, ws.next()).await {
                Err(_) => return Err(Error::AnswerTimeout("no usage reply within timeout".into())),
                Ok(None) | Ok(Some(Ok(Message::Close(_)))) | Ok(Some(Err(_))) => {
                    return Err(Error::ManagerUnavailable("connection closed".into()))
                }
                Ok(Some(Ok(Message::Text(t)))) => t,
                Ok(Some(Ok(Message::Binary(b)))) => String::from_utf8_lossy(&b).into_owned(),
                Ok(Some(Ok(_))) => continue,
            };
            let msg: serde_json::Value = serde_json::from_str(&frame)
                .map_err(|_| Error::Protocol(format!("manager sent non-JSON frame: {frame}")))?;
            match msg.get("type").and_then(|t| t.as_str()) {
                Some("usage") => return Ok(msg),
                Some("err") => {
                    return Err(Error::Protocol(
                        msg.get("text").and_then(|v| v.as_str()).unwrap_or("usage error").to_string(),
                    ))
                }
                _ => continue, // skip ack / a / initialized
            }
        }
    }

    async fn fetch_token(&self) -> Result<String> {
        let p = self.token_provider.clone();
        tokio::task::spawn_blocking(move || p())
            .await
            .map_err(|e| Error::Auth(format!("token task failed: {e}")))?
    }

    /// Open the WebSocket and drain the manager's `initialized` frame.
    pub async fn connect(&mut self) -> Result<()> {
        let token = self.fetch_token().await?;
        let mut request = self
            .url
            .as_str()
            .into_client_request()
            .map_err(|e| Error::ManagerUnavailable(format!("bad manager URL {}: {e}", self.url)))?;
        request.headers_mut().insert(
            "Authorization",
            format!("Bearer {token}")
                .parse()
                .map_err(|e| Error::ManagerUnavailable(format!("bad auth header: {e}")))?,
        );
        let config = WebSocketConfig {
            max_message_size: None, // Python sets max_size=None
            max_frame_size: None,
            ..Default::default()
        };
        let connect = connect_async_with_config(request, Some(config), false);
        let (ws, _resp) = match tokio::time::timeout(Duration::from_secs(15), connect).await {
            Err(_) => {
                return Err(Error::ManagerUnavailable(format!(
                    "could not connect to {}: open timed out",
                    self.url
                )))
            }
            Ok(Err(e)) => {
                return Err(Error::ManagerUnavailable(format!(
                    "could not connect to {}: {e}",
                    self.url
                )))
            }
            Ok(Ok(pair)) => pair,
        };
        self.ws = Some(ws);

        // The manager leads with an `initialized` frame; capture the sid.
        match tokio::time::timeout(Duration::from_secs(10), self.recv_text()).await {
            Ok(Ok(Some(text))) => {
                if let Ok(v) = serde_json::from_str::<serde_json::Value>(&text) {
                    if v.get("type").and_then(|t| t.as_str()) == Some("initialized") {
                        self.session_id =
                            v.get("sid").and_then(|s| s.as_str()).map(String::from);
                        tracing::debug!(
                            "connected sid={:?} backendPort={:?}",
                            self.session_id,
                            v.get("backendPort")
                        );
                    } else {
                        tracing::warn!(
                            "expected 'initialized', got {:?}",
                            v.get("type")
                        );
                    }
                }
                Ok(())
            }
            _ => {
                self.close().await;
                Err(Error::ManagerUnavailable(format!(
                    "no 'initialized' frame from {}",
                    self.url
                )))
            }
        }
    }

    pub async fn close(&mut self) {
        if let Some(mut ws) = self.ws.take() {
            let _ = ws.close(None).await;
        }
    }

    /// Receive the next text-ish message, or None if the socket closed/erred.
    async fn recv_text(&mut self) -> Result<Option<String>> {
        let ws = match self.ws.as_mut() {
            Some(w) => w,
            None => return Ok(None),
        };
        loop {
            match ws.next().await {
                Some(Ok(Message::Text(t))) => return Ok(Some(t)),
                Some(Ok(Message::Binary(b))) => {
                    return Ok(Some(String::from_utf8_lossy(&b).into_owned()))
                }
                Some(Ok(Message::Close(_))) | None => return Ok(None),
                Some(Ok(_)) => continue, // ping/pong/frame — keep reading
                Some(Err(_)) => return Ok(None),
            }
        }
    }

    /// Send a question and return its settled `Answer`. Transparently reconnects
    /// (up to `max_reconnects`) if the socket dropped.
    pub async fn ask(&mut self, text: &str, timeout: Duration) -> Result<Answer> {
        let mut attempts = 0u32;
        loop {
            if !self.connected() {
                self.connect().await?;
            }
            self.counter += 1;
            let msg_id = format!("m{}", self.counter);
            let q = json!({"type": "q", "id": msg_id, "text": text}).to_string();

            if let Some(ws) = self.ws.as_mut() {
                if ws.send(Message::Text(q)).await.is_err() {
                    self.close().await;
                    attempts += 1;
                    if attempts > self.max_reconnects {
                        return Err(Error::ManagerUnavailable(
                            "connection closed during ask()".into(),
                        ));
                    }
                    tracing::warn!("connection dropped, reconnecting (attempt {attempts})");
                    continue;
                }
            }

            match self.await_answer(&msg_id, Instant::now() + timeout).await? {
                Awaited::Answer(a) => return Ok(a),
                Awaited::Closed => {
                    self.close().await;
                    attempts += 1;
                    if attempts > self.max_reconnects {
                        return Err(Error::ManagerUnavailable(
                            "connection closed during ask()".into(),
                        ));
                    }
                    tracing::warn!("connection dropped, reconnecting (attempt {attempts})");
                    continue;
                }
            }
        }
    }

    async fn await_answer(&mut self, msg_id: &str, deadline: Instant) -> Result<Awaited> {
        loop {
            let remaining = deadline.saturating_duration_since(Instant::now());
            if remaining.is_zero() {
                return Err(Error::AnswerTimeout(format!(
                    "no answer for {msg_id} within the timeout"
                )));
            }
            // Scope the &mut borrow so we can re-borrow self.ws to send `confirm`.
            let frame = {
                let ws = match self.ws.as_mut() {
                    Some(w) => w,
                    None => return Ok(Awaited::Closed),
                };
                match tokio::time::timeout(remaining, ws.next()).await {
                    Err(_) => {
                        return Err(Error::AnswerTimeout(format!(
                            "no answer for {msg_id} within the timeout"
                        )))
                    }
                    Ok(None) | Ok(Some(Ok(Message::Close(_)))) | Ok(Some(Err(_))) => {
                        return Ok(Awaited::Closed)
                    }
                    Ok(Some(Ok(Message::Text(t)))) => t,
                    Ok(Some(Ok(Message::Binary(b)))) => String::from_utf8_lossy(&b).into_owned(),
                    Ok(Some(Ok(_))) => continue, // ping/pong
                }
            };

            let msg: serde_json::Value = serde_json::from_str(&frame)
                .map_err(|_| Error::Protocol(format!("manager sent non-JSON frame: {frame}")))?;
            match msg.get("type").and_then(|t| t.as_str()) {
                Some("a") if msg.get("id").and_then(|i| i.as_str()) == Some(msg_id) => {
                    let seq = msg.get("seq").and_then(|s| s.as_i64()).unwrap_or(-1);
                    let confirm = json!({"type": "confirm", "seq": seq}).to_string();
                    if let Some(ws) = self.ws.as_mut() {
                        let _ = ws.send(Message::Text(confirm)).await;
                    }
                    return Ok(Awaited::Answer(Answer {
                        text: msg.get("text").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                        seq,
                        id: msg
                            .get("id")
                            .and_then(|v| v.as_str())
                            .unwrap_or(msg_id)
                            .to_string(),
                        time_in: msg.get("timeIn").and_then(|v| v.as_str()).map(String::from),
                        time_out: msg.get("timeOut").and_then(|v| v.as_str()).map(String::from),
                        latency_ms: msg.get("latencyMs").and_then(|v| v.as_i64()),
                    }));
                }
                Some("err") => {
                    return Err(Error::Protocol(
                        msg.get("text")
                            .and_then(|v| v.as_str())
                            .unwrap_or("manager returned an error")
                            .to_string(),
                    ))
                }
                // `initialized`, `ack`, or an unrelated `a` → keep waiting.
                other => {
                    tracing::debug!("skip frame type={:?} id={:?}", other, msg.get("id"));
                }
            }
        }
    }
}
